File: src/main/java/com/odeyalo/sonata/authorization/repository/BasicPersistentOperations.java
package com.odeyalo.sonata.authorization.repository;

import org.jetbrains.annotations.NotNull;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

/**
 * Represents the basic operations that can be done using persistent storage
 */
public interface BasicPersistentOperations<T, ID> {

    @NotNull
    Mono<T> findById(@NotNull ID id);

    @NotNull
    Flux<T> findAll();

    @NotNull
    Mono<T> save(@NotNull T entity);

    @NotNull
    Mono<Void> deleteById(@NotNull ID id);

}

File: src/main/java/com/odeyalo/sonata/authorization/repository/BasicPersistentOperations.kt
package com.odeyalo.sonata.authorization.repository

import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

/**
 * Represent the basic operations that can be done using persistent storage
 */
interface BasicPersistentOperations<T, ID> {

    fun findById(id: ID): Mono<T>;

    fun findAll(): Flux<T>

    fun save(entity: T): Mono<T>

    fun deleteById(id: ID): Mono<Void>

}